### ISSUE TITLE
Graceful shutdown is not working for async server

### ISSUE THREAD

**Issue by IlianIliev** [grpc/grpc#26123]
### What version of gRPC and what language are you using?
1.37

### What operating system (Linux, Windows,...) and version?
macOS Cataline 10.15.7

### What runtime / compiler are you using (e.g. python version or version of gcc)
Python 3.9.0

### What did you do?
I am running the following code
```
import logging
import asyncio
import grpc


async def serve() -> None:
    server = grpc.aio.server()
    listen_addr = "[::]:50051"
    server.add_insecure_port(listen_addr)
    logging.info("Starting server on %s", listen_addr)
    await server.start()
    try:
        await server.wait_for_termination()
    except KeyboardInterrupt:
        # Shuts down the server with 0 seconds of grace period. During the
        # grace period, the server won't accept new connections and allow
        # existing RPCs to continue within the grace period.
        await server.stop(0)


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(serve())
```

### What did you expect to see?
A graceful shutdown

### What did you see instead?
```
/Users/ilian/.virtualenvs/usersettings-service/bin/python /Users/ilian/workspace/lifesum/usersettings-service/tmp.py
INFO:root:Starting server on [::]:50051
Traceback (most recent call last):
  File "/Users/ilian/workspace/lifesum/usersettings-service/tmp.py", line 38, in <module>
    asyncio.run(serve())
  File "/Users/ilian/.pyenv/versions/3.9.0/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/ilian/.pyenv/versions/3.9.0/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/Users/ilian/.pyenv/versions/3.9.0/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/Users/ilian/.pyenv/versions/3.9.0/lib/python3.9/asyncio/base_events.py", line 1854, in _run_once
    event_list = self._selector.select(timeout)
  File "/Users/ilian/.pyenv/versions/3.9.0/lib/python3.9/selectors.py", line 562, in select
    kev_list = self._selector.control(None, max_ev, timeout)
KeyboardInterrupt
Exception ignored in: <function Server.__del__ at 0x10b05c5e0>
Traceback (most recent call last):
  File "/Users/ilian/.virtualenvs/usersettings-service/lib/python3.9/site-packages/grpc/aio/_server.py", line 169, in __del__
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 116, in grpc._cython.cygrpc.schedule_coro_threadsafe
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/common.pyx.pxi", line 108, in grpc._cython.cygrpc.schedule_coro_threadsafe
  File "/Users/ilian/.pyenv/versions/3.9.0/lib/python3.9/asyncio/base_events.py", line 431, in create_task
  File "/Users/ilian/.pyenv/versions/3.9.0/lib/python3.9/asyncio/base_events.py", line 510, in _check_closed
RuntimeError: Event loop is closed
sys:1: RuntimeWarning: coroutine 'AioServer.shutdown' was never awaited
```

**Comment by damarvin** [comment]
Same under Windows, Python 3.9.5, grpcio-tools==1.38.1.
I need to give CTRL+C twice to actually stop the server.
No success with some exception handling trials.

**Comment by lidizheng** [comment]
Thanks for the report. With latest `v1.38.1`, I setup a gRPC client / gRPC server. The gRPC server method takes 2 seconds to finish. The graceful shutdown suppose to wait for the existing RPC to finish, but it didn't:

```
$ python3 async_greeter_server.py
INFO:root:Starting server on [::]:50051
^CTraceback (most recent call last):
  File "async_greeter_server.py", line 50, in <module>
    asyncio.run(serve())
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 574, in run_until_complete
    self.run_forever()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 541, in run_forever
    self._run_once()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 1750, in _run_once
    event_list = self._selector.select(timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/selectors.py", line 558, in select
    kev_list = self._selector.control(None, max_ev, timeout)
KeyboardInterrupt
```

This is definitely something I want to improve.

---

@damarvin What you described seems like another issue. It could be the first CTRL+C makes the server enter graceful state, and the second CTRL+C force exit the graceful state and end the application. Can you file a separate issue with reproduce snippets?

**Comment by damarvin** [comment]
I see exactly the output as the case is about, with the 1st CTRL+C already.
I assume @IlianIliev needs a 2nd too, to get back on the console, so I'd wait for #26123 is fixed.

**Comment by lidizheng** [comment]
I created PR https://github.com/grpc/grpc/pull/26622 to add a proper graceful shutdown example. @IlianIliev  Please let me know if it works or not.

@damarvin The 2 CTRL+C problem is different than the graceful shutdown described in this issue. The graceful shutdown issue is about where can `KeyInterrupt` be propagated to. Can you share your reproduce snippet?

**Comment by IlianIliev** [comment]
@lidizheng it works for me, thanks a lot for the help.

**Comment by lidizheng** [comment]
Glad to hear that!

**Comment by IlianIliev** [comment]
@lidizheng unfortunately, I found another related issue. I explain it here, and please let me know if I should create a separate ticket for it.

In the example, you created -> https://github.com/grpc/grpc/pull/26622/files everything works fine.
However if you replace the `asyncio.sleep` call at https://github.com/grpc/grpc/pull/26622/files#diff-a71e80ec63a3fbeff2bdf255101f67e3b77d794424851f153ca4afa271c8ee6bR33 with a normal sleep, the request is no longer completed.
Example
```
    async def SayHello(
            self, request: helloworld_pb2.HelloRequest,
            context: grpc.aio.ServicerContext) -> helloworld_pb2.HelloReply:
        logging.info('Received request, sleeping for 4 seconds...')
        # await asyncio.sleep(4)
        import time
        time.sleep(4)
        logging.info('Sleep completed, responding')
        return helloworld_pb2.HelloReply(message='Hello, %s!' % request.name)
```

Stopping the server during the request will result in the following exception in the client:
```
Traceback (most recent call last):
  File "/Users/ilian/workspace/grpc/examples/python/helloworld/async_greeter_client.py", line 33, in <module>
    asyncio.run(run())
  File "/Users/ilian/.pyenv/versions/3.9.0/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/ilian/.pyenv/versions/3.9.0/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/Users/ilian/workspace/grpc/examples/python/helloworld/async_greeter_client.py", line 27, in run
    response = await stub.SayHello(helloworld_pb2.HelloRequest(name='you'))
  File "/Users/ilian/.virtualenvs/grpc/lib/python3.9/site-packages/grpc/aio/_call.py", line 285, in __await__
    raise _create_rpc_error(self._cython_call._initial_metadata,
grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Socket closed"
	debug_error_string = "{"created":"@1626950326.447489000","description":"Error received from peer ipv6:[::1]:50051","file":"src/core/lib/surface/call.cc","file_line":1070,"grpc_message":"Socket closed","grpc_status":14}"
>
```

In addition, if you increase the shutdown time for the server to 10 seconds, and time the requests you will notice the following:
 - with async-sleep, the request completes in ~4 seconds, and then the server shuts down immediately
 - with sync-sleep, the duration of the failing request is ~10 seconds. So it actually just sits and waits for the server to die, but the call is never completed


Do you have any idea what may be causing it.

**Comment by lidizheng** [comment]
With async-sleep, the coroutine yield the right of execution back to the event loop, which is the desired behavior. With sync-sleep, it blocks the entire thread, so no progress can be made on any other coroutines. In another word, if the server method handler uses sync-sleep, the shutdown process is stalled.

Within asyncio's realm, all blocking calls need to be replaced with asyncio version ones to avoid the problem you mentioned. Like synchronization locks, data structures to https://docs.python.org/3/library/asyncio-sync.html. Also, file I/O.

**Comment by IlianIliev** [comment]
I assumed something similar, but I can not agree that this is a normal behaviour.
In my test scenario I use 5 second sync sleep, with 10 seconds for shutdown. This way the server has 5 extra seconds to complete the request, which does not happen.

In addition, what happens if we are running some slower calculation or anything time-consuming that is not I/O. Even if we talk about something that takes 100ms there is still a chance that the shutdown happens during this time and the task will be killed again, without any chance for completing.

**Comment by lidizheng** [comment]
> In my test scenario I use 5 second sync sleep, with 10 seconds for shutdown. This way the server has 5 extra seconds to complete the request, which does not happen.

It's possible that when the server enters sync sleep, the shutdown timer hasn't been created.

> what happens if we are running some slower calculation or anything time-consuming that is not I/O.

This is the weak spot of single-threaded asynchronous programming. For Java or C#, this generally won't be a problem, because they have a thread pool to schedule all the tasks. For Node.js or Python, if I anticipate a very compute-heavy task (like image processing, video encoding, ml stuff), I would try to manage it as a separate process [asyncio.subprocess](https://docs.python.org/3/library/asyncio-subprocess.html) or `multiprocess`.

> Even if we talk about something that takes 100ms there is still a chance that the shutdown happens during this time and the task will be killed again, without any chance for completing.

Computational-heavy tasks won't kill other coroutines. It merely blocks the queue of execution, which means it will stall coroutines from running concurrently.

**Comment by IlianIliev** [comment]
> It's possible that when the server enters sync sleep, the shutdown timer hasn't been created.
Even if this is the case, once the sync-sleep is done, with the 5/10 scenario we have 5 more seconds to complete the request and return a response.

> This is the weak spot of single-threaded asynchronous programming...
Can you please help me understand a bit how it works under the hood. My assumption is that each gRPC call is run as a separate coroutine. So, when the shutdown is initiated, we stop receiving requests, try to gather the results from all coroutines, and if something is not ready in the shutdown time it is killed.

> Computational-heavy tasks won't kill other coroutines. It merely blocks the queue of execution, which means it will stall coroutines from running concurrently.
But wouldn't that block the queue in the same way as the sync-sleep so the method will never complete. Which practically makes the graceful shutdown not so graceful.

**Comment by lidizheng** [comment]
I personally found the official asyncio develop guide quite helpful: https://docs.python.org/3/library/asyncio-dev.html#running-blocking-code

**Comment by IlianIliev** [comment]
That part I understand. But regarding the graceful shutdown, the only difference between running a CPU-intensive call for 1s, and something for 1ms is that the second case has a lower probability of happening at the same time as the shutdown call.

**Comment by lidizheng** [comment]
I tried your change. If we use `time.sleep` the `KeyboardInterrupt` exception will be populated to the `time.sleep`, hence it will not be caught by our new try-catch clause at all.

**Comment by IlianIliev** [comment]
It should be caught as I see that the "INFO:root:Starting graceful shutdown..." when I try to stop the server during the `time.sleep`.

**Comment by IlianIliev** [comment]
I played a bit more by myself, and I think that now I understand better what you mean. If you look at this example -> https://gist.github.com/IlianIliev/9aba04a74a4faddf0749533205d3b001

When we use `async.sleep` all works fine, and we can complete a graceful shutdown, but with `time.sleep` it does not, as the `KeyboardInterrupt` is raised from the sleep call which makes the function impossible to continue.
I also tried this with looping over a large list and the result was the same, the exception is raised from the loop.

But the results from above mean, that if we have a highly loaded gRPC server handling multiple requests simultaneously, the chance of the shutdown killing one of those requests is quite high. Potentially wrapping each request handler in a try/except block that can catch the exception and let the rest of the handlers complete could improve the situation, but we will still have failed handlers. Not fun (

**Comment by lidizheng** [comment]
Good point. I reproduced your result with a busy loop in the server handler. In normal Python, `KeyboardInterrupt` will be populated to the main thread, so it's deterministic. But with asyncio, the main thread might be occupied by the request handler or the event loop, hence the `KeyboardInterrupt` problem exists.

This sounds like a design problem for `asyncio`. I don't have an answer. Maybe we should open an issue to CPython? https://bugs.python.org/

**Comment by IlianIliev** [comment]
I have opened a ticket there -> https://bugs.python.org/msg398400
Let's see if they can help in any way. And thank you for your help to look at this issue )
